python/pyarrow/types.pxi

```diff
@@ -114,7 +123,14 @@ cdef void* _as_c_pointer(v, allow_null=False) except *:
             "Arrow library", UserWarning, stacklevel=2)
         c_ptr = <void*> <uintptr_t > v
     elif PyCapsule_CheckExact(v):
-        c_ptr = PyCapsule_GetPointer(v, NULL)
+        capsule_name = PyCapsule_GetName(v)
+        if capsule_name == NULL or strcmp(capsule_name, "r_extptr") == 0:
```

Comment:
We have a `frombytes` helper that you could also use here `frombytes(capsule_name) == "r_extptr`.

Cython generally converts between python and c objects transparently (in which case `capsule_name == b"r_extptr"` might work out of the box), although in the case of a C string (`char*`) it might be that an explicit assignment is needed from reading the docs.

Reply:
("transparently" was maybe the wrong word choice ;))